# Practice Lab for Types of Data

You are given a record of dogs that have visited a small vet clinic in the past few hours. The clinic is currently very busy. There is a new virus that turns some dogs (who are usually by default all good boys and girls) into bad doggies. Also a very nasty doggy flu has been going around at the same time. 

* Given the situation, data entry has been rushed and there are some inconsistencies that need to be addressed and type conversion you will need to do to get the data into your organization's standard. 
* Your direct manager (who is not a data person) is European and wants the dog length to be converted to metric because he thinks that the inch is a preposterous unit of measurement.
* The head veterinarian wants a flagging system for dogs that come in running a fever.


You, the only data analyst at this small clinic, have been assigned these tasks. Exercise your knowledge of string cleaning, mathematical operations on numeric quantities, and booleans in order to complete the work.


In [25]:
# load data given to you
dog_data = [['Name', 'Breed', 'Good Doggy?', 'Length [in]', 'Weight Range', 'Body Temp [F]'],
    ["Buddy", "Golden Retriever", "Good", 24, 'Overweight', 101.5],
    ["Cooper", "Doberman Pinscher", "Bad", 28, 'Normal',  105.2],
    ["Max", "GERMAN Shepherd", "Good", 26, 'Normal', 100.2],
    ["Bella", "Chow Chow", "Bad", 20, 'Normal', 100.5],
    ["Lucy", " Yorkshire Terrier", "Good", 9, 'Normal', 102.0],
    ["Charlie", "Beagle", "Good", 13, 'Overweight', 100.4],
    ["Daisy", "Akita", "Bad", 24, 'Normal',99.8],
    ["Bailey", "JACK russell terrier", "Bad", 10, 'Underweight', 101.2],
    ["Lola", "Rottweiler   ", "Good", 27, 'Normal', 102.5],
    ["Sadie", "Siberian_Husky", "Bad", 23, 'Normal', 99.5],
    ["Penny", "STANDARD Poodle", "Good", 18, 'Underweight', 100.7],
    ["Bruno", "Bulldog", "Bad", 14, 'Normal', 104.1],
    ["Rosie", "Beagle", "Good", 13, 'Normal', 100.6],
    ["Duke", "    DOBERMAN PINSCHER", "Good", 28,  'Normal',100.1],
    ["Luna", "Shiba Inu", "Good", 17, 'Obese',  100.3],
    ["Scout", "Golden_Retriever", "Good", 22, 'Normal', 103.5],
    ["Rex", "Rottweiler", "Bad", 23, 'Normal', 100.2],
]

### Instructions

1. Write a clean_string function that takes in a string and removes leading/trailing spaces, replaces underscores with spaces, and capitalizes the first letter of each word. Use this function to replace the values in the 'Breed' column with the new cleaned values.

In [26]:
def clean_string(s):
    return s.strip().title().replace(' ', '_')

# clean the data
for row in dog_data[1:]: # skips the header
    row[1] = clean_string(row[1])
    


In [27]:
dog_data

[['Name',
  'Breed',
  'Good Doggy?',
  'Length [in]',
  'Weight Range',
  'Body Temp [F]'],
 ['Buddy', 'Golden_Retriever', 'Good', 24, 'Overweight', 101.5],
 ['Cooper', 'Doberman_Pinscher', 'Bad', 28, 'Normal', 105.2],
 ['Max', 'German_Shepherd', 'Good', 26, 'Normal', 100.2],
 ['Bella', 'Chow_Chow', 'Bad', 20, 'Normal', 100.5],
 ['Lucy', 'Yorkshire_Terrier', 'Good', 9, 'Normal', 102.0],
 ['Charlie', 'Beagle', 'Good', 13, 'Overweight', 100.4],
 ['Daisy', 'Akita', 'Bad', 24, 'Normal', 99.8],
 ['Bailey', 'Jack_Russell_Terrier', 'Bad', 10, 'Underweight', 101.2],
 ['Lola', 'Rottweiler', 'Good', 27, 'Normal', 102.5],
 ['Sadie', 'Siberian_Husky', 'Bad', 23, 'Normal', 99.5],
 ['Penny', 'Standard_Poodle', 'Good', 18, 'Underweight', 100.7],
 ['Bruno', 'Bulldog', 'Bad', 14, 'Normal', 104.1],
 ['Rosie', 'Beagle', 'Good', 13, 'Normal', 100.6],
 ['Duke', 'Doberman_Pinscher', 'Good', 28, 'Normal', 100.1],
 ['Luna', 'Shiba_Inu', 'Good', 17, 'Obese', 100.3],
 ['Scout', 'Golden_Retriever', 'Good', 22, 

2. Transform lengths from inches to centimeters.

In [28]:
for row in dog_data[1:]:
    row[3] = round(row[3] * 2.54, 2)
dog_data[0][3] = 'Lenght[cm]'

In [29]:
dog_data

[['Name',
  'Breed',
  'Good Doggy?',
  'Lenght[cm]',
  'Weight Range',
  'Body Temp [F]'],
 ['Buddy', 'Golden_Retriever', 'Good', 60.96, 'Overweight', 101.5],
 ['Cooper', 'Doberman_Pinscher', 'Bad', 71.12, 'Normal', 105.2],
 ['Max', 'German_Shepherd', 'Good', 66.04, 'Normal', 100.2],
 ['Bella', 'Chow_Chow', 'Bad', 50.8, 'Normal', 100.5],
 ['Lucy', 'Yorkshire_Terrier', 'Good', 22.86, 'Normal', 102.0],
 ['Charlie', 'Beagle', 'Good', 33.02, 'Overweight', 100.4],
 ['Daisy', 'Akita', 'Bad', 60.96, 'Normal', 99.8],
 ['Bailey', 'Jack_Russell_Terrier', 'Bad', 25.4, 'Underweight', 101.2],
 ['Lola', 'Rottweiler', 'Good', 68.58, 'Normal', 102.5],
 ['Sadie', 'Siberian_Husky', 'Bad', 58.42, 'Normal', 99.5],
 ['Penny', 'Standard_Poodle', 'Good', 45.72, 'Underweight', 100.7],
 ['Bruno', 'Bulldog', 'Bad', 35.56, 'Normal', 104.1],
 ['Rosie', 'Beagle', 'Good', 33.02, 'Normal', 100.6],
 ['Duke', 'Doberman_Pinscher', 'Good', 71.12, 'Normal', 100.1],
 ['Luna', 'Shiba_Inu', 'Good', 43.18, 'Obese', 100.3],


3. Convert Good/Bad doggy column to boolean

In [30]:
for row in dog_data[1:]:
   row[2] = row[2] == 'Good'

In [31]:
dog_data

[['Name',
  'Breed',
  'Good Doggy?',
  'Lenght[cm]',
  'Weight Range',
  'Body Temp [F]'],
 ['Buddy', 'Golden_Retriever', True, 60.96, 'Overweight', 101.5],
 ['Cooper', 'Doberman_Pinscher', False, 71.12, 'Normal', 105.2],
 ['Max', 'German_Shepherd', True, 66.04, 'Normal', 100.2],
 ['Bella', 'Chow_Chow', False, 50.8, 'Normal', 100.5],
 ['Lucy', 'Yorkshire_Terrier', True, 22.86, 'Normal', 102.0],
 ['Charlie', 'Beagle', True, 33.02, 'Overweight', 100.4],
 ['Daisy', 'Akita', False, 60.96, 'Normal', 99.8],
 ['Bailey', 'Jack_Russell_Terrier', False, 25.4, 'Underweight', 101.2],
 ['Lola', 'Rottweiler', True, 68.58, 'Normal', 102.5],
 ['Sadie', 'Siberian_Husky', False, 58.42, 'Normal', 99.5],
 ['Penny', 'Standard_Poodle', True, 45.72, 'Underweight', 100.7],
 ['Bruno', 'Bulldog', False, 35.56, 'Normal', 104.1],
 ['Rosie', 'Beagle', True, 33.02, 'Normal', 100.6],
 ['Duke', 'Doberman_Pinscher', True, 71.12, 'Normal', 100.1],
 ['Luna', 'Shiba_Inu', True, 43.18, 'Obese', 100.3],
 ['Scout', 'Golden

4. Convert the ‘Weight’ column to an integer representation.

5. Create a new column called Has_Fever. The value should yield 0 if a dog has normal body temperature and 1 if a dog is running a fever. A dog should be flagged as having a fever if their body temperature is above 102 F.

6. Run cell to display all changes you have made.

In [ ]:
dog_data